### ISSUE TITLE
ODBC bulk copy operations request

### ISSUE THREAD

**Issue by keitherskine** [mkleehammer/pyodbc#350]
This isn't an issue as such, more of a request.  It's a bit specific to SQL Server, but are there any plans to implement the ODBC bulk copy functions described here: https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-extensions-bulk-copy-functions/sql-server-driver-extensions-bulk-copy-functions

If I can do `bcp` style operations through pyodbc rather than through the command line, that would be very helpful for me.

**Comment by boumboum** [comment]
Have you tried the fast_executemany feature?
This is not a bcp but the speed is quite decent.

**Comment by gordthompson** [comment]
I was curious to see if there was a significant difference in speed between the two approaches, so I did a quick informal test using a CSV file with 100,000 rows ...

```
1,record000000
2,record000001
3,record000002
...
100000,record099999
```

... inserting into an empty table on a remote SQL Server instance.

The times for `bcp` to do the insert were highly variable, ranging from 7 seconds to 32 seconds.

For `fast_executemany` to do the insert, i.e.,

```python
cnxn = pyodbc.connect(conn_str, autocommit=False)
crsr = cnxn.cursor()
crsr.fast_executemany = True
t0 = time.time()
data = []
with open(r'C:\Users\Gord\Desktop\test10k.csv', newline='') as csvfile:
    rdr = csv.reader(csvfile, delimiter=',', quoting=csv.QUOTE_NONE)
    data = [(int(row[0]), row[1]) for row in rdr]
sql = 'INSERT INTO test10k (id, txt) VALUES (?,?)'
crsr.executemany(sql, data)
cnxn.commit()
print(f"{(time.time() - t0):.1f} seconds")
```

the times were more consistent, ranging from 20 seconds to 30 seconds.

So, `fast_executemany` may turn out to be slower than `bcp`, but perhaps not outrageously so.

**Comment by keitherskine** [comment]
Thank you for running some tests @gordthompson .  In my case, I am importing gigabytes, if not tens of gigabytes, of data on a regular basis so holding all that in memory would be problematic.  Yes, I could split the data into chunks but still, it's a lot simpler to just load the data directly from disc file to database.  Also, taking the best times from your tests, the `fast_executemany` appears to be 3 times slower than `bcp`.  That's not bad, but is still quite a considerable difference.

Right now, I run `bcp` commands through python using `subprocess` which does work of course, but if running bulk loads through pyodbc was available I would almost certainly use it.

**Comment by nrhy** [comment]
I vaguely recall there being language specific libraries from different database vendors - something like sybase's perl bcp library. If i'm not mistaken, it allowed bcp calls using in memory data, contrary to the bcp command line utility. Is this no longer a thing?

**Comment by damionw** [comment]
I'm a bit late to the party, but, in case anyone still cares, we wrote this around 2008 and have been using it in production. It was open sourced around 2013

https://github.com/damionw/pythonbcp

**Comment by bkline** [comment]
Hi, @keitherskine. I've got a [pull request](https://github.com/mkleehammer/pyodbc/pull/1486) in the hopper. I applied the [80/20 rule](https://en.wikipedia.org/wiki/Pareto_principle) here, providing what I hope will meet the needs of most users. It's implemented as a `Connection` method. Simplest usage looks like this (though you'd want to turn on `autocommit` for a real import):

<img width="1049" height="449" alt="Image" src="https://github.com/user-attachments/assets/84e687ed-40df-4d38-a309-45fee83fbd31" />

But you can also provide custom format files, process a slice of the rows, and control what happens with errors:

```
Help on built-in function bcp:

bcp(...) method of pyodbc.Connection instance
    bcp(self, action, table, datafile, /, **kwargs)

    Required positional arguments
      action: BCP_IN or BCP_OUT
      table: required name of table to copy or populate
      datafile: path to table data file

    Optional keyword arguments
      formatfile: custom specification for copying
      errorfile: used to report rows which failed import
      firstrow: used to skip past the first rows (1-based)
      lastrow: tells bcp to stop after processing the row at this position
      maxerrors: tells bcp to abort after detecting this many errors (default 10)
```

Some notes on the implementation.
- Dynamic location of the `bcp_*` functions required a different technique for each of the various platforms (Windows, Linux, macOS). I don't have access to a box on which to test BSD, but I expect it probably works the same as macOS, which is based on BSD.
- The FreeTDS driver doesn't include those functions, which live in a different library (libsybdb). I did some experimenting with letting the user tell us where that library can be found, but when I started hitting segfaults I gave it up. I'm not prepared to say it's impossible to support FreeTDS `bcp`, but I'm leaving that support out for now.
- There's reasonably good test coverage. I focused on making sure my own code was doing the right thing, rather than trying to also flush out potential bugs in Microsoft's handling of the various data types. So, for example, my test tables have four columns, unlike the FreeTDS test suite, which has sixteen.
- Unfortunately, although Microsoft bundles and documents these as ODBC extension functions, they aren't very closely coupled with the rest of ODBC. Most notably, when a `bcp_*` function returns `FAIL` there's no way to get diagnostic records with details about what went wrong. I can tell you which function failed, but that's about it. If you catch a `pyodbc.OperationalError` exception saying "bcp_init failure" that might mean there was a typo in the table name, or an input file which could not be read, or a garbled format file, or ... (you get the idea).

Taking into account your comment about the amount of data with which you work, I scaled up Gord's benchmark by four orders of magnitude, importing from a data file with a billion rows. Using `cursor.executemany()` with `fast_executemany` set took a bit under five hours to load the data into a table, giving a throughput of roughly 57,300 rows per second.. Once you get above the threshold where the bulk copy option is worthwhile (somewhere under 10,00 rows of the size we're using for Gord's test) the performance improves by an order of magnitude by comparison with `fast_executemany`. Here are some numbers (not very scientifically achieved, as I didn't go to any trouble to eliminate caching).

### Import from CSV datafile

Number of Rows | Elapsed Time | Total Seconds | Rows Per Second
---: | --- | ---: | ---:
1 | 0:00:00.023780 | 0.02 | 42.1
10 | 0:00:00.196429 | 0.20 | 50.9
100 | 0:00:00.242034 | 0.24 | 413.2
1,000 | 0:00:00.233693 | 0.23 | 4,279.1
10,000 | 0:00:00.085836 | 0.09 | 116,501.2
100,000 | 0:00:00.217588 | 0.22 | 459,584.2
1,000,000 | 0:00:01.972877 | 1.97 | 506,874.0
10,000,000 | 0:00:21.364803 | 21.36 | 468,059.5
100,000,000 | 0:03:37.583316 | 217.58 | 459,594.1
1,000,000,000 | 0:52:56.149866 | 3,176.15 | 314,846.6

You can see that the curve starts dropping significantly somewhere after 100 million rows, with the knee of the curve around one million rows.

Those runs kept the original benchmark's approach of using a character-based datafile. I was curious to see how the default native format would compare. The output numbers are breathtaking, peaking at just under 3 million rows per second (I didn't time similar export runs using the character format). It turned out that the performance for import from a native data file does not differ dramatically from character mode.

### Native export

Number of Rows | Elapsed Time | Total Seconds | Rows Per Second
---: | --- | ---: | ---:
1 | 0:00:00.045688 | 0.02 | 21.9
10 | 0:00:00.006976 | 0.01 | 1,433.5
100 | 0:00:00.007883 | 0.01 | 12,685.5
1,000 | 0:00:00.010841 | 0.01 | 92,242.4
10,000 | 0:00:00.021660 | 0.02 | 461,680.5
100,000 | 0:00:00.080338 | 0.08 | 1,244,741.0
1,000,000 | 0:00:00.360306 | 0.36 | 2,775,418.7
10,000,000 | 0:00:03.799016 | 3.80 | 2,632,260.6
100,000,000 | 0:00:34.281851 | 34.3 | 2,916,995.4
1,000,000,000 | 0:05:50.949290 | 351.0 | 2,849,414.5

### Native import

Number of Rows | Elapsed Time | Total Seconds | Rows Per Second
---: | --- | ---: | ---:
1 | 0:00:00.025192 | 0.03 | 39.7
10 | 0:00:00.013578 | 0.02 | 736.5
100 | 0:00:00.014931 | 0.01 | 6,697.5
1,000 | 0:00:00.022996 | 0.02 | 43,485.8
10,000 | 0:00:00.076276 | 0.08 | 131,102.8
100,000 | 0:00:00.213454 | 0.21 | 468,485.0
1,000,000 | 0:00:01.931533 | 1.93 | 517,723.5
10,000,000 | 0:00:29.951930 | 29.95 | 333,868.3
100,000,000 | 0:03:37.709257 | 217.71 | 459,328.2
1,000,000,000 | 0:51:21.233630 | 3,081.23 | 324,545.3

## TL;DR

BCP is fast.

---
Let me know if this resembles in any way what you were hoping for.
